File: db-tester-junit/src/main/java/io/github/seijikohara/dbtester/junit/jupiter/extension/DatabaseTest.java
package io.github.seijikohara.dbtester.junit.jupiter.extension;

import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Inherited;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import org.junit.jupiter.api.extension.ExtendWith;

/**
 * Enables database testing support for a JUnit Jupiter test class.
 *
 * <p>This is a composed annotation that combines {@link ExtendWith
 * ExtendWith(DatabaseTestExtension.class)}, reducing boilerplate for test authors.
 *
 * <p><b>Usage:</b>
 *
 * <pre>{@code
 * @DatabaseTest
 * class UserRepositoryTest {
 *
 *     @BeforeAll
 *     static void setUp(ExtensionContext context) {
 *         DataSource dataSource = createDataSource();
 *         DatabaseTestExtension.getRegistry(context).registerDefault(dataSource);
 *     }
 *
 *     @Test
 *     @DataSet
 *     @ExpectedDataSet
 *     void shouldVerifyDatabaseState() {
 *         // test implementation
 *     }
 * }
 * }</pre>
 *
 * <p>This annotation is {@link Inherited}, so subclasses of annotated test classes automatically
 * inherit the database testing extension.
 *
 * @see DatabaseTestExtension
 * @see io.github.seijikohara.dbtester.api.annotation.DataSet
 * @see io.github.seijikohara.dbtester.api.annotation.ExpectedDataSet
 */
@Target(ElementType.TYPE)
@Retention(RetentionPolicy.RUNTIME)
@Inherited
@Documented
@ExtendWith(DatabaseTestExtension.class)
public @interface DatabaseTest {}

File: db-tester-kotest/src/main/kotlin/io/github/seijikohara/dbtester/kotest/annotation/DatabaseTest.kt
package io.github.seijikohara.dbtester.kotest.annotation

import io.github.seijikohara.dbtester.kotest.extension.DatabaseTestExtension
import io.github.seijikohara.dbtester.kotest.extension.DatabaseTestSupport
import io.kotest.core.extensions.ApplyExtension

/**
 * Annotation that enables database testing for a Kotest spec class.
 *
 * When applied to a spec class, this annotation automatically registers
 * the [DatabaseTestExtension] for handling [DataSet][io.github.seijikohara.dbtester.api.annotation.DataSet]
 * and [ExpectedDataSet][io.github.seijikohara.dbtester.api.annotation.ExpectedDataSet] annotations.
 *
 * **Important:** The spec class must implement [DatabaseTestSupport] to provide
 * the [DataSourceRegistry][io.github.seijikohara.dbtester.api.config.DataSourceRegistry].
 *
 * **Usage:**
 * ```kotlin
 * @DatabaseTest
 * class MySpec : AnnotationSpec(), DatabaseTestSupport {
 *
 *     override val dbTesterRegistry = DataSourceRegistry()
 *
 *     @BeforeAll
 *     fun setup() {
 *         dbTesterRegistry.registerDefault(createDataSource())
 *     }
 *
 *     @Test
 *     @DataSet
 *     @ExpectedDataSet
 *     fun `should verify database state`() {
 *         // test implementation
 *     }
 * }
 * ```
 *
 * **With custom configuration:**
 * ```kotlin
 * @DatabaseTest
 * class MySpec : AnnotationSpec(), DatabaseTestSupport {
 *
 *     override val dbTesterRegistry = DataSourceRegistry()
 *
 *     override val dbTesterConfiguration = Configuration.builder()
 *         .conventions(ConventionSettings.builder()
 *             .dataFormat(DataFormat.TSV)
 *             .build())
 *         .build()
 *
 *     @Test
 *     @DataSet
 *     fun `should load TSV data`() {
 *         // test implementation
 *     }
 * }
 * ```
 *
 * @see DatabaseTestExtension
 * @see DatabaseTestSupport
 * @see io.github.seijikohara.dbtester.api.annotation.DataSet
 * @see io.github.seijikohara.dbtester.api.annotation.ExpectedDataSet
 */
@Target(AnnotationTarget.CLASS)
@Retention(AnnotationRetention.RUNTIME)
@MustBeDocumented
@ApplyExtension(DatabaseTestExtension::class)
annotation class DatabaseTest
